In [ ]:
!pip install -U "flwr[simulation]" torch==2.8.0 opacus matplotlib

In [ ]:
# =========================
# Federated Healthcare (Colab)
# Flower (simulation) + PyTorch
# =========================

import json, math, random, warnings
from dataclasses import dataclass
from typing import Dict, List, Tuple

import numpy as np
import pandas as pd


from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import (
    accuracy_score, precision_recall_fscore_support, roc_auc_score,
    classification_report, confusion_matrix
)
from sklearn.impute import SimpleImputer

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset

import flwr as fl
from flwr.common import (
    ndarrays_to_parameters,
    parameters_to_ndarrays,
    NDArrays,
    Scalar,
)

warnings.filterwarnings("ignore")

# -----------------------------
# Config & Utilities
# -----------------------------

@dataclass
class FLConfig:
    num_clients: int = 8
    num_rounds: int = 5
    local_epochs: int = 1
    batch_size: int = 32
    lr: float = 1e-3
    seed: int = 42
    dirichlet_alpha: float = 0.5  # client heterogeneity
    dp_on: bool = False           # optional DP
    dp_noise_multiplier: float = 1.0
    dp_max_grad_norm: float = 1.0


def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)


def is_binary_labels(y: np.ndarray) -> bool:
    return len(np.unique(y)) == 2


# -----------------------------
# Load & preprocess
# -----------------------------

def engineer_features(df: pd.DataFrame) -> pd.DataFrame:
    """Create useful numeric features from dates and amounts."""
    df = df.copy()

    # Length of stay (days) if dates exist
    if "Date of Admission" in df.columns and "Discharge Date" in df.columns:
        adm = pd.to_datetime(df["Date of Admission"], errors="coerce")
        dis = pd.to_datetime(df["Discharge Date"], errors="coerce")
        df["length_of_stay_days"] = (dis - adm).dt.days

    # Clean billing
    if "Billing Amount" in df.columns:
        df["Billing Amount"] = pd.to_numeric(df["Billing Amount"], errors="coerce")

    # Normalize casing for certain categoricals
    for col in ["Gender", "Blood Type", "Medical Condition", "Admission Type",
                "Medication", "Insurance Provider"]:
        if col in df.columns and df[col].dtype == object:
            df[col] = df[col].astype(str).str.strip().str.title()

    return df


def load_healthcare(csv_path: str, target_col: str = "stroke"):
    """
    Loads stroke dataset, drops obvious ID/PII, preprocesses features,
    and returns train/test splits suitable for FL.
    """
    df = pd.read_csv(csv_path)

    # If the stroke column is not present, raise an error
    if target_col not in df.columns:
        raise ValueError(f"Target column '{target_col}' not found. Available: {list(df.columns)}")

    # Drop obvious non-predictive identifiers if they exist
    drop_cols = [c for c in ["id", "Name", "Doctor", "Hospital"] if c in df.columns]
    df = df.drop(columns=drop_cols, errors="ignore")

    # Feature engineering (won't do much here but safe to keep)
    df = engineer_features(df)

    # Target: stroke (0/1)
    y_raw = df[target_col]
    # Force to numeric 0/1
    y = pd.to_numeric(y_raw, errors="coerce").astype(float)
    # Drop rows where target is NaN after conversion
    mask = ~np.isnan(y)
    df = df.loc[mask].reset_index(drop=True)
    y = y[mask].astype(int).values

    # Features: all columns except target
    X = df.drop(columns=[target_col])

    numeric_cols = X.select_dtypes(include=[np.number]).columns.tolist()
    categorical_cols = [c for c in X.columns if c not in numeric_cols]

    preprocessor = ColumnTransformer(
        transformers=[
            ("num", Pipeline([
                ("imputer", SimpleImputer(strategy="median")),
                ("scaler", StandardScaler())
            ]), numeric_cols),
            ("cat", Pipeline([
                ("imputer", SimpleImputer(strategy="most_frequent")),
                ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
            ]), categorical_cols),
        ]
    )

    strat = y if is_binary_labels(y) else None
    X_train_raw, X_test_raw, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=strat
    )

    X_train = preprocessor.fit_transform(X_train_raw)
    X_test = preprocessor.transform(X_test_raw)

    X_train = np.array(X_train)
    X_test = np.array(X_test)
    y_train = np.array(y_train)
    y_test = np.array(y_test)

    return X_train, X_test, y_train, y_test, preprocessor



# -----------------------------
# Federated partition (non-IID)
# -----------------------------

def dirichlet_partition(X, y, num_clients: int, alpha: float = 0.5, seed: int = 42):
    """Non-IID Dirichlet partition of data into num_clients splits."""
    rng = np.random.default_rng(seed)
    classes = np.unique(y)
    idx_by_class = {c: np.where(y == c)[0] for c in classes}
    client_indices = [[] for _ in range(num_clients)]

    for c in classes:
        idxs = idx_by_class[c]
        rng.shuffle(idxs)
        props = rng.dirichlet(alpha=[alpha] * num_clients)
        counts = np.floor(props * len(idxs)).astype(int)
        while counts.sum() < len(idxs):
            counts[rng.integers(0, num_clients)] += 1
        start = 0
        for i in range(num_clients):
            end = start + counts[i]
            client_indices[i].extend(idxs[start:end].tolist())
            start = end

    splits = []
    X = np.array(X)
    y = np.array(y)
    for ci in client_indices:
        ci = np.array(ci, dtype=int)
        # Only include splits with data
        if len(ci) > 0:
            splits.append((X[ci], y[ci]))
    return splits


# -----------------------------
# Flatten helpers for parameters
# -----------------------------

def flatten_ndarrays(nds: List[np.ndarray]) -> Tuple[np.ndarray, List[Tuple[int, ...]]]:
    """Flatten a list of ndarrays into a single 1D vector + remember shapes."""
    shapes = [a.shape for a in nds]
    flats = [a.ravel() for a in nds]
    flat = np.concatenate(flats).astype(np.float64)
    return flat, shapes

def unflatten_ndarrays(flat: np.ndarray, shapes: List[Tuple[int, ...]]) -> List[np.ndarray]:
    """Rebuild list of ndarrays from flat vector + shapes."""
    out = []
    i = 0
    for s in shapes:
        n = int(np.prod(s))
        part = flat[i:i+n].reshape(s)
        out.append(part)
        i += n
    return out


# -----------------------------
# Model & Training helpers
# -----------------------------

class MLP(nn.Module):
    def __init__(self, in_dim: int):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(in_dim, 128), nn.ReLU(),
            nn.Linear(128, 64), nn.ReLU(),
            nn.Linear(64, 1)
        )

    def forward(self, x):
        return self.net(x).squeeze(1)


def make_pos_weight(y: np.ndarray):
    """For BCEWithLogitsLoss: pos_weight = N_neg / N_pos."""
    unique, counts = np.unique(y, return_counts=True)
    cdict = dict(zip(unique, counts))
    if 0 in cdict and 1 in cdict and cdict[1] > 0:
        return torch.tensor(cdict[0] / cdict[1], dtype=torch.float32)
    return torch.tensor(1.0, dtype=torch.float32)


def to_tensor_dataset(X: np.ndarray, y: np.ndarray) -> TensorDataset:
    return TensorDataset(torch.tensor(X, dtype=torch.float32),
                         torch.tensor(y, dtype=torch.float32))


def bce_metrics(logits: np.ndarray, y_true: np.ndarray) -> Dict[str, float]:
    probs = 1 / (1 + np.exp(-logits))
    y_pred = (probs >= 0.5).astype(int)
    acc = accuracy_score(y_true, y_pred)
    prec, rec, f1, _ = precision_recall_fscore_support(
        y_true, y_pred, average="binary", zero_division=0
    )
    try:
        auc = roc_auc_score(y_true, probs)
    except Exception:
        auc = float("nan")
    return {"accuracy": acc, "precision": prec, "recall": rec, "f1": f1, "roc_auc": auc}


# -----------------------------
# Flower Client
# -----------------------------

class TabularClient(fl.client.NumPyClient):
    def __init__(self, cid: int, X: np.ndarray, y: np.ndarray, input_dim: int, cfg: FLConfig):
        self.cid = cid
        self.X = X
        self.y = y
        self.cfg = cfg

        self.model = MLP(input_dim)

        if is_binary_labels(self.y):
            self.criterion = nn.BCEWithLogitsLoss(pos_weight=make_pos_weight(self.y))
        else:
            self.criterion = nn.BCEWithLogitsLoss()

        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=self.cfg.lr)

    def get_parameters(self, config={}):
        return [v.cpu().numpy() for _, v in self.model.state_dict().items()]

    def set_parameters(self, params):
        state_dict = self.model.state_dict()
        for (k, _), v in zip(state_dict.items(), params):
            state_dict[k] = torch.tensor(v)
        self.model.load_state_dict(state_dict)

    def fit(self, params, config={}):
        self.set_parameters(params)
        # Ensure X is not empty before creating DataLoader
        if len(self.X) == 0:
            print(f"Client {self.cid} has no data, skipping fit.")
            return self.get_parameters(), 0, {}

        loader = DataLoader(
            to_tensor_dataset(self.X, self.y),
            batch_size=self.cfg.batch_size,
            shuffle=True,
        )
        self.model.train()
        for _ in range(self.cfg.local_epochs):
            for xb, yb in loader:
                logits = self.model(xb)
                loss = self.criterion(logits, yb)
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()
        return self.get_parameters(), len(self.X), {}

    def evaluate(self, params, config={}):
        self.set_parameters(params)
        if len(self.X) == 0:
            print(f"Client {self.cid} has no data, skipping evaluation.")
            return 0.0, 0, {"accuracy": 0.0, "precision": 0.0, "recall": 0.0, "f1": 0.0, "roc_auc": float("nan")}

        self.model.eval()
        with torch.no_grad():
            logits = self.model(torch.tensor(self.X, dtype=torch.float32)).cpu().numpy()
        m = bce_metrics(logits, self.y)
        loss = float(
            nn.BCEWithLogitsLoss()(
                torch.tensor(logits, dtype=torch.float32),
                torch.tensor(self.y, dtype=torch.float32)
            ).item()
        )
        return loss, len(self.X), m


# -----------------------------
# Server-side (test set) evaluation
# -----------------------------

def gen_evaluate_fn(X_test: np.ndarray, y_test: np.ndarray, input_dim: int, cfg: FLConfig):
    def evaluate(server_round: int, parameters: fl.common.NDArrays, config):
        model = MLP(input_dim)
        state_dict = model.state_dict()
        for (k, _), v in zip(state_dict.items(), parameters):
            state_dict[k] = torch.tensor(v)
        model.load_state_dict(state_dict)

        model.eval()
        with torch.no_grad():
            logits = model(torch.tensor(X_test, dtype=torch.float32)).cpu().numpy()
        metrics = bce_metrics(logits, y_test)

        print(f"[Round {server_round}] test: " +
              json.dumps({k: round(v, 4) if v == v else None for k, v in metrics.items()}))

        loss = float(
            nn.BCEWithLogitsLoss()(
                torch.tensor(logits, dtype=torch.float32),
                torch.tensor(y_test, dtype=torch.float32)
            ).item()
        )
        return loss, metrics
    return evaluate


# -----------------------------
# Custom FedAvg that logs updates for HE notebook
# -----------------------------

class LoggingFedAvg(fl.server.strategy.FedAvg):
    """
    Same as FedAvg, but on a chosen round it logs client deltas Δw_i
    (local_i - global_after_round) to 'round1_updates.npy'
    for later use in a separate HE notebook.
    """
    def __init__(self, log_round: int = 1, log_path: str = "round1_updates.npy", **kwargs):
        super().__init__(**kwargs)
        self.log_round = log_round
        self.log_path = log_path
        self._shapes_cache = None
        self._logged = False

    def aggregate_fit(
        self,
        rnd: int,
        results: List[Tuple[fl.server.client_proxy.ClientProxy, fl.server.client_proxy.FitRes]],
        failures: List[BaseException],
    ) -> Tuple[NDArrays, Dict[str, Scalar]]:

        # 1) Let FedAvg do the usual aggregation
        aggregated_params, metrics = super().aggregate_fit(rnd, results, failures)

        # 2) Log once, for the chosen round
        if (not self._logged) and (rnd == self.log_round) and results:
            print(f"📥 Logging client updates from round {rnd} to '{self.log_path}'")

            # Global AFTER aggregation for this round
            global_nd = parameters_to_ndarrays(aggregated_params)
            flat_global, shapes = flatten_ndarrays(global_nd)
            self._shapes_cache = shapes

            updates = []
            for _, fitres in results:
                local_nd = parameters_to_ndarrays(fitres.parameters)  # client's local model
                flat_local, _ = flatten_ndarrays(local_nd)
                delta = flat_local - flat_global  # Δw_i = local_i - global_after
                updates.append(delta)

            updates = np.stack(updates, axis=0)  # shape: [num_clients, D]

            np.save(self.log_path, updates)
            np.save(self.log_path.replace(".npy", "_shapes.npy"),
                    np.array(self._shapes_cache, dtype=object))
            print(f"✅ Saved shape {updates.shape} to '{self.log_path}'")

            self._logged = True

        return aggregated_params, metrics



# -----------------------------
# Orchestration
# -----------------------------

def run_federated(csv_path: str, target_col: str = "Test Results", cfg: FLConfig = FLConfig()):
    set_seed(cfg.seed)

    # Load & preprocess
    X_train, X_test, y_train, y_test, preproc = load_healthcare(csv_path, target_col)
    input_dim = X_train.shape[1]

    # Non-IID client splits
    client_splits = dirichlet_partition(
        X_train, y_train, cfg.num_clients,
        alpha=cfg.dirichlet_alpha, seed=cfg.seed
    )

    def client_fn(cid: str):
        i = int(cid)
        # Ensure client_splits has enough elements
        if i < len(client_splits):
            Xc, yc = client_splits[i]
            return TabularClient(i, Xc, yc, input_dim, cfg)
        else:
            # Handle cases where client_splits might have fewer clients than cfg.num_clients
            # This can happen if some partitions ended up empty and were filtered out.
            # For now, we return a dummy client that doesn't train/evaluate.
            print(f"Client {cid} requested, but no data available. Returning a dummy client.")
            return TabularClient(i, np.array([]).reshape(0, input_dim), np.array([]), input_dim, cfg)

    strategy = LoggingFedAvg(
        log_round=1,                     # log round-1 updates for HE notebook
        log_path="round1_updates.npy",
        evaluate_fn=gen_evaluate_fn(X_test, y_test, input_dim, cfg),
        fraction_fit=1.0,
        fraction_evaluate=1.0,
        min_fit_clients=cfg.num_clients, # Changed this to allow simulation to proceed with fewer actual clients
        min_evaluate_clients=cfg.num_clients,
        min_available_clients=cfg.num_clients,
    )

    hist = fl.simulation.start_simulation(
        client_fn=client_fn,
        num_clients=cfg.num_clients,
        config=fl.server.ServerConfig(num_rounds=cfg.num_rounds),
        strategy=strategy,
    )

    print("\n✅ Training completed successfully.")
    return hist, X_train, X_test, y_test


# -----------------------------
# Run
# -----------------------------

hist, X_train, X_test, y_test = run_federated(
    "healthcare-dataset-stroke-data.csv",
    target_col="stroke",
    cfg=FLConfig(num_clients=8, num_rounds=5, local_epochs=1, batch_size=32, lr=1e-3),
)

print("✅ Training complete.")

	Instead, use the `flwr run` CLI command to start a local simulation in your Flower app, as shown for example below:

		$ flwr new  # Create a new Flower app from a template

		$ flwr run  # Run the Flower app in Simulation Mode

	Using `start_simulation()` is deprecated.

            This is a deprecated feature. It will be removed
            entirely in future versions of Flower.
        
INFO :      Starting Flower simulation, config: num_rounds=5, no round_timeout
2025-12-08 22:03:54,589	INFO worker.py:2012 -- Started a local Ray instance.
INFO :      Flower VCE: Ray initialized with resources: {'memory': 9174541108.0, 'CPU': 2.0, 'object_store_memory': 3931946188.0, 'node:__internal_head__': 1.0, 'node:172.28.0.12': 1.0}
INFO :      Optimize your simulation with Flower VCE: https://flower.ai/docs/framework/how-to-run-simulations.html
INFO :      No `client_resources` specified. Using minimal resources for clients.
INFO :      Flower VCE: Resources for each Virtual Client: {'num_c

[Round 0] test: {"accuracy": 0.0489, "precision": 0.0489, "recall": 1.0, "f1": 0.0933, "roc_auc": 0.3437}


(ClientAppActor pid=8232) WARNING :   DEPRECATED FEATURE: `client_fn` now expects a signature `def client_fn(context: Context)`.The provided `client_fn` has signature: {'cid': <Parameter "cid: str">}. You can import the `Context` like this: `from flwr.common import Context`
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)             This is a deprecated feature. It will be removed
(ClientAppActor pid=8232)             entirely in future versions of Flower.
(ClientAppActor pid=8232)         
(ClientAppActor pid=8232) WARNING :   Deprecation Warning: The `client_fn` function must return an instance of `Client`, but an instance of `NumpyClient` was returned. Please use `NumPyClient.to_client()` method to convert it to `Client`.
(ClientAppActor pid=8232) WARNING :   DEPRECATED FEATURE: `client_fn` now expects a signature `def client_fn(context: Context)`.The provided `client_fn` has signature: {'cid': <Parameter "cid: str">}. You can import the `Context` like this: `from flwr.common i

📥 Logging client updates from round 1 to 'round1_updates.npy'
✅ Saved shape (8, 11137) to 'round1_updates.npy'
[Round 1] test: {"accuracy": 0.7886, "precision": 0.1513, "recall": 0.72, "f1": 0.25, "roc_auc": 0.8143}


(ClientAppActor pid=8233) WARNING :   DEPRECATED FEATURE: `client_fn` now expects a signature `def client_fn(context: Context)`.The provided `client_fn` has signature: {'cid': <Parameter "cid: str">}. You can import the `Context` like this: `from flwr.common import Context`
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)             This is a deprecated feature. It will be removed
(ClientAppActor pid=8233)             entirely in future versions of Flower.
(ClientAppActor pid=8233)         
(ClientAppActor pid=8233) WARNING :   Deprecation Warning: The `client_fn` function must return an instance of `Client`, but an instance of `NumpyClient` was returned. Please use `NumPyClient.to_client()` method to convert it to `Client`.
(ClientAppActor pid=8233) /usr/local/lib/python3.12/dist-packages/sklearn/metrics/_ranking.py:379: UndefinedMetricWarning: Only one class is present in y_true. ROC AUC score is not defined in that case.
(ClientAppActor pid=8233)   warnings.warn(
(ClientAppActo

[Round 2] test: {"accuracy": 0.8327, "precision": 0.173, "recall": 0.64, "f1": 0.2723, "roc_auc": 0.8237}


(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 3]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(

[Round 3] test: {"accuracy": 0.7661, "precision": 0.1434, "recall": 0.76, "f1": 0.2413, "roc_auc": 0.825}


(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 4]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
INFO :      aggregate_fit: received 8 results and 0 failures
(ClientAppActor pid=8233) 
(C

[Round 4] test: {"accuracy": 0.7632, "precision": 0.1391, "recall": 0.74, "f1": 0.2342, "roc_auc": 0.8222}


(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 5]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(

[Round 5] test: {"accuracy": 0.7671, "precision": 0.1412, "recall": 0.74, "f1": 0.2372, "roc_auc": 0.8236}


(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
(ClientAppActor pid=8233) 
(ClientAppActor pid=8233)         
(ClientAppActor pid=8232) 
(ClientAppActor pid=8232)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [SUMMARY]
INFO :      Run finished 5 round(s) in 11.15s
INFO :      	History (loss, distributed):
INFO :      		round 1: 0.5904511463356344
INFO :      		round 2: 0.4621336572163842
INFO :      		round 3: 0.4368923114076869
INFO :      		round 4: 0.42530070630188327
INFO :      		round 5: 0.4120056913134053
INFO :      	History (loss, centralized):
INFO :      		round 0: 0.7512249946594238
INFO :      		round 1: 0.5859930515289307
INFO :      		round 2: 0.45264530181884766
INFO :      		round 3: 0.423700213432312
INFO :      		roun


✅ Training completed successfully.
✅ Training complete.


In [ ]:
display(hist.metrics_centralized)

{'accuracy': [(0, 0.04892367906066536),
  (1, 0.7886497064579256),
  (2, 0.8326810176125244),
  (3, 0.7661448140900196),
  (4, 0.7632093933463796),
  (5, 0.7671232876712328)],
 'precision': [(0, 0.04892367906066536),
  (1, 0.15126050420168066),
  (2, 0.17297297297297298),
  (3, 0.14339622641509434),
  (4, 0.13909774436090225),
  (5, 0.14122137404580154)],
 'recall': [(0, 1.0), (1, 0.72), (2, 0.64), (3, 0.76), (4, 0.74), (5, 0.74)],
 'f1': [(0, 0.09328358208955224),
  (1, 0.25),
  (2, 0.2723404255319149),
  (3, 0.24126984126984127),
  (4, 0.23417721518987342),
  (5, 0.23717948717948717)],
 'roc_auc': [(0, np.float64(0.34374485596707816)),
  (1, np.float64(0.8142798353909465)),
  (2, np.float64(0.8237037037037037)),
  (3, np.float64(0.8249588477366254)),
  (4, np.float64(0.8221810699588477)),
  (5, np.float64(0.8235596707818931))]}

In [ ]:
# 1. Mount Drive
from google.colab import drive
drive.mount('/content/drive')

# 2. Load updates saved by FL
import numpy as np
updates = np.load("/content/round1_updates.npy")
print("Loaded updates from FL:", updates.shape)

# 3. Blockchain import
import sys
sys.path.append('/content/drive/MyDrive')
from mock_ledger import MockBlockchain

ledger = MockBlockchain()

# Store each client update
for i in range(updates.shape[0]):
    payload = updates[i].tobytes()
    ledger.submit_update(1, f"client_{i}", payload)

# 4. Save ledger locally
import pickle
with open("ledger.pkl", "wb") as f:
    pickle.dump(ledger, f)

print("Saved ledger to local file 'ledger.pkl'")

#Copy all files into Drive
!cp round1_updates.npy /content/drive/MyDrive/
!cp round1_updates_shapes.npy /content/drive/MyDrive/
!cp ledger.pkl /content/drive/MyDrive/

print(" All files copied into Google Drive")


Mounted at /content/drive
Loaded updates from FL: (8, 11137)
[Ledger] Stored update: round=1, client=client_0, hash=647694ff...
[Ledger] Stored update: round=1, client=client_1, hash=3bbe88b3...
[Ledger] Stored update: round=1, client=client_2, hash=036089d1...
[Ledger] Stored update: round=1, client=client_3, hash=0cb6653f...
[Ledger] Stored update: round=1, client=client_4, hash=fcf3b545...
[Ledger] Stored update: round=1, client=client_5, hash=a9340d24...
[Ledger] Stored update: round=1, client=client_6, hash=0ff11e16...
[Ledger] Stored update: round=1, client=client_7, hash=fd3a4a08...
Saved ledger to local file 'ledger.pkl'
 All files copied into Google Drive


# Task
Install `imbalanced-learn`, then modify the `load_healthcare` function to apply SMOTE to the training data after preprocessing and before federated partitioning. Update the `FLConfig` and `bce_metrics` to use a `eval_threshold` of `0.7` for predictions. Finally, run the federated learning simulation, display the centralized metrics, and interpret the impact of SMOTE and the higher `eval_threshold` on precision, recall, and other metrics.

## Install imbalanced-learn

### Subtask:
Install the `imbalanced-learn` library, which provides the SMOTE algorithm, in the Colab environment.


**Reasoning**:
To install the `imbalanced-learn` library, I will use the `!pip install` command.



In [ ]:
!pip install imbalanced-learn

**Reasoning**:
The next step is to modify the `load_healthcare` function to apply SMOTE to the training data after preprocessing and before federated partitioning, as per the main task. This involves importing SMOTE and integrating it into the data loading and preprocessing pipeline.



In [ ]:
import json, math, random, warnings
from dataclasses import dataclass
from typing import Dict, List, Tuple

import numpy as np
import pandas as pd


from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import (
    accuracy_score, precision_recall_fscore_support, roc_auc_score,
    classification_report, confusion_matrix
)
from sklearn.impute import SimpleImputer
from imblearn.over_sampling import SMOTE # Import SMOTE

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset

import flwr as fl
from flwr.common import (
    ndarrays_to_parameters,
    parameters_to_ndarrays,
    NDArrays,
    Scalar,
)

warnings.filterwarnings("ignore")

# -----------------------------
# Config & Utilities
# -----------------------------

@dataclass
class FLConfig:
    num_clients: int = 8
    num_rounds: int = 5
    local_epochs: int = 1
    batch_size: int = 32
    lr: float = 1e-3
    seed: int = 42
    dirichlet_alpha: float = 0.5  # client heterogeneity
    dp_on: bool = False           # optional DP
    dp_noise_multiplier: float = 1.0
    dp_max_grad_norm: float = 1.0
    eval_threshold: float = 0.5   # New threshold for predictions


def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)


def is_binary_labels(y: np.ndarray) -> bool:
    return len(np.unique(y)) == 2


# -----------------------------
# Load & preprocess
# -----------------------------

def engineer_features(df: pd.DataFrame) -> pd.DataFrame:
    """Create useful numeric features from dates and amounts."""
    df = df.copy()

    # Length of stay (days) if dates exist
    if "Date of Admission" in df.columns and "Discharge Date" in df.columns:
        adm = pd.to_datetime(df["Date of Admission"], errors="coerce")
        dis = pd.to_datetime(df["Discharge Date"], errors="coerce")
        df["length_of_stay_days"] = (dis - adm).dt.days

    # Clean billing
    if "Billing Amount" in df.columns:
        df["Billing Amount"] = pd.to_numeric(df["Billing Amount"], errors="coerce")

    # Normalize casing for certain categoricals
    for col in ["Gender", "Blood Type", "Medical Condition", "Admission Type",
                "Medication", "Insurance Provider"]:
        if col in df.columns and df[col].dtype == object:
            df[col] = df[col].astype(str).str.strip().str.title()

    return df


def load_healthcare(csv_path: str, target_col: str = "stroke", apply_smote: bool = False):
    """
    Loads stroke dataset, drops obvious ID/PII, preprocesses features,
    and returns train/test splits suitable for FL.
    """
    df = pd.read_csv(csv_path)

    # If the stroke column is not present, raise an error
    if target_col not in df.columns:
        raise ValueError(f"Target column '{target_col}' not found. Available: {list(df.columns)}")

    # Drop obvious non-predictive identifiers if they exist
    drop_cols = [c for c in ["id", "Name", "Doctor", "Hospital"] if c in df.columns]
    df = df.drop(columns=drop_cols, errors="ignore")

    # Feature engineering (won't do much here but safe to keep)
    df = engineer_features(df)

    # Target: stroke (0/1)
    y_raw = df[target_col]
    # Force to numeric 0/1
    y = pd.to_numeric(y_raw, errors="coerce").astype(float)
    # Drop rows where target is NaN after conversion
    mask = ~np.isnan(y)
    df = df.loc[mask].reset_index(drop=True)
    y = y[mask].astype(int).values

    # Features: all columns except target
    X = df.drop(columns=[target_col])

    numeric_cols = X.select_dtypes(include=[np.number]).columns.tolist()
    categorical_cols = [c for c in X.columns if c not in numeric_cols]

    preprocessor = ColumnTransformer(
        transformers=[
            ("num", Pipeline([
                ("imputer", SimpleImputer(strategy="median")),
                ("scaler", StandardScaler())
            ]), numeric_cols),
            ("cat", Pipeline([
                ("imputer", SimpleImputer(strategy="most_frequent")),
                ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
            ]), categorical_cols),
        ]
    )

    strat = y if is_binary_labels(y) else None
    X_train_raw, X_test_raw, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=strat
    )

    X_train = preprocessor.fit_transform(X_train_raw)
    X_test = preprocessor.transform(X_test_raw)

    # Apply SMOTE if requested, after preprocessing
    if apply_smote:
        print("Applying SMOTE to training data...")
        smote = SMOTE(random_state=42)
        X_train, y_train = smote.fit_resample(X_train, y_train)
        print(f"Training data shape after SMOTE: X={X_train.shape}, y={y_train.shape}")

    X_train = np.array(X_train)
    X_test = np.array(X_test)
    y_train = np.array(y_train)
    y_test = np.array(y_test)

    return X_train, X_test, y_train, y_test, preprocessor



# -----------------------------
# Federated partition (non-IID)
# -----------------------------

def dirichlet_partition(X, y, num_clients: int, alpha: float = 0.5, seed: int = 42):
    """Non-IID Dirichlet partition of data into num_clients splits."""
    rng = np.random.default_rng(seed)
    classes = np.unique(y)
    idx_by_class = {c: np.where(y == c)[0] for c in classes}
    client_indices = [[] for _ in range(num_clients)]

    for c in classes:
        idxs = idx_by_class[c]
        rng.shuffle(idxs)
        props = rng.dirichlet(alpha=[alpha] * num_clients)
        counts = np.floor(props * len(idxs)).astype(int)
        while counts.sum() < len(idxs):
            counts[rng.integers(0, num_clients)] += 1
        start = 0
        for i in range(num_clients):
            end = start + counts[i]
            client_indices[i].extend(idxs[start:end].tolist())
            start = end

    splits = []
    X = np.array(X)
    y = np.array(y)
    for ci in client_indices:
        ci = np.array(ci, dtype=int)
        # Only include splits with data
        if len(ci) > 0:
            splits.append((X[ci], y[ci]))
    return splits


# -----------------------------
# Flatten helpers for parameters
# -----------------------------

def flatten_ndarrays(nds: List[np.ndarray]) -> Tuple[np.ndarray, List[Tuple[int, ...]]]:
    """Flatten a list of ndarrays into a single 1D vector + remember shapes."""
    shapes = [a.shape for a in nds]
    flats = [a.ravel() for a in nds]
    flat = np.concatenate(flats).astype(np.float64)
    return flat, shapes

def unflatten_ndarrays(flat: np.ndarray, shapes: List[Tuple[int, ...]]) -> List[np.ndarray]:
    """Rebuild list of ndarrays from flat vector + shapes."""
    out = []
    i = 0
    for s in shapes:
        n = int(np.prod(s))
        part = flat[i:i+n].reshape(s)
        out.append(part)
        i += n
    return out


# -----------------------------
# Model & Training helpers
# -----------------------------

class MLP(nn.Module):
    def __init__(self, in_dim: int):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(in_dim, 128), nn.ReLU(),
            nn.Linear(128, 64), nn.ReLU(),
            nn.Linear(64, 1)
        )

    def forward(self, x):
        return self.net(x).squeeze(1)


def make_pos_weight(y: np.ndarray):
    """For BCEWithLogitsLoss: pos_weight = N_neg / N_pos."""
    unique, counts = np.unique(y, return_counts=True)
    cdict = dict(zip(unique, counts))
    if 0 in cdict and 1 in cdict and cdict[1] > 0:
        return torch.tensor(cdict[0] / cdict[1], dtype=torch.float32)
    return torch.tensor(1.0, dtype=torch.float32)


def to_tensor_dataset(X: np.ndarray, y: np.ndarray) -> TensorDataset:
    return TensorDataset(torch.tensor(X, dtype=torch.float32),
                         torch.tensor(y, dtype=torch.float32))


def bce_metrics(logits: np.ndarray, y_true: np.ndarray, eval_threshold: float = 0.5) -> Dict[str, float]:
    probs = 1 / (1 + np.exp(-logits))
    y_pred = (probs >= eval_threshold).astype(int) # Use eval_threshold here
    acc = accuracy_score(y_true, y_pred)
    prec, rec, f1, _ = precision_recall_fscore_support(
        y_true, y_pred, average="binary", zero_division=0
    )
    try:
        auc = roc_auc_score(y_true, probs)
    except Exception:
        auc = float("nan")
    return {"accuracy": acc, "precision": prec, "recall": rec, "f1": f1, "roc_auc": auc}


# -----------------------------
# Flower Client
# -----------------------------

class TabularClient(fl.client.NumPyClient):
    def __init__(self, cid: int, X: np.ndarray, y: np.ndarray, input_dim: int, cfg: FLConfig):
        self.cid = cid
        self.X = X
        self.y = y
        self.cfg = cfg

        self.model = MLP(input_dim)

        if is_binary_labels(self.y):
            self.criterion = nn.BCEWithLogitsLoss(pos_weight=make_pos_weight(self.y))
        else:
            self.criterion = nn.BCEWithLogitsLoss()

        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=self.cfg.lr)

    def get_parameters(self, config={}):
        return [v.cpu().numpy() for _, v in self.model.state_dict().items()]

    def set_parameters(self, params):
        state_dict = self.model.state_dict()
        for (k, _), v in zip(state_dict.items(), params):
            state_dict[k] = torch.tensor(v)
        self.model.load_state_dict(state_dict)

    def fit(self, params, config={}):
        self.set_parameters(params)
        # Ensure X is not empty before creating DataLoader
        if len(self.X) == 0:
            print(f"Client {self.cid} has no data, skipping fit.")
            return self.get_parameters(), 0, {}

        loader = DataLoader(
            to_tensor_dataset(self.X, self.y),
            batch_size=self.cfg.batch_size,
            shuffle=True,
        )
        self.model.train()
        for _ in range(self.cfg.local_epochs):
            for xb, yb in loader:
                logits = self.model(xb)
                loss = self.criterion(logits, yb)
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()
        return self.get_parameters(), len(self.X), {}

    def evaluate(self, params, config={}):
        self.set_parameters(params)
        if len(self.X) == 0:
            print(f"Client {self.cid} has no data, skipping evaluation.")
            return 0.0, 0, {"accuracy": 0.0, "precision": 0.0, "recall": 0.0, "f1": 0.0, "roc_auc": float("nan")}

        self.model.eval()
        with torch.no_grad():
            logits = self.model(torch.tensor(self.X, dtype=torch.float32)).cpu().numpy()
        m = bce_metrics(logits, self.y, eval_threshold=self.cfg.eval_threshold) # Pass eval_threshold
        loss = float(
            nn.BCEWithLogitsLoss()(
                torch.tensor(logits, dtype=torch.float32),
                torch.tensor(self.y, dtype=torch.float32)
            ).item()
        )
        return loss, len(self.X), m


# -----------------------------
# Server-side (test set) evaluation
# -----------------------------

def gen_evaluate_fn(X_test: np.ndarray, y_test: np.ndarray, input_dim: int, cfg: FLConfig):
    def evaluate(server_round: int, parameters: fl.common.NDArrays, config):
        model = MLP(input_dim)
        state_dict = model.state_dict()
        for (k, _), v in zip(state_dict.items(), parameters):
            state_dict[k] = torch.tensor(v)
        model.load_state_dict(state_dict)

        model.eval()
        with torch.no_grad():
            logits = model(torch.tensor(X_test, dtype=torch.float32)).cpu().numpy()
        metrics = bce_metrics(logits, y_test, eval_threshold=cfg.eval_threshold) # Pass eval_threshold

        print(f"[Round {server_round}] test: " +
              json.dumps({k: round(v, 4) if v == v else None for k, v in metrics.items()}))

        loss = float(
            nn.BCEWithLogitsLoss()(
                torch.tensor(logits, dtype=torch.float32),
                torch.tensor(y_test, dtype=torch.float32)
            ).item()
        )
        return loss, metrics
    return evaluate


# -----------------------------
# Custom FedAvg that logs updates for HE notebook
# -----------------------------

class LoggingFedAvg(fl.server.strategy.FedAvg):
    """
    Same as FedAvg, but on a chosen round it logs client deltas Δw_i
    (local_i - global_after_round) to 'round1_updates.npy'
    for later use in a separate HE notebook.
    """
    def __init__(self, log_round: int = 1, log_path: str = "round1_updates.npy", **kwargs):
        super().__init__(**kwargs)
        self.log_round = log_round
        self.log_path = log_path
        self._shapes_cache = None
        self._logged = False

    def aggregate_fit(
        self,
        rnd: int,
        results: List[Tuple[fl.server.client_proxy.ClientProxy, fl.server.client_proxy.FitRes]],
        failures: List[BaseException],
    ) -> Tuple[NDArrays, Dict[str, Scalar]]:

        # 1) Let FedAvg do the usual aggregation
        aggregated_params, metrics = super().aggregate_fit(rnd, results, failures)

        # 2) Log once, for the chosen round
        if (not self._logged) and (rnd == self.log_round) and results:
            print(f"📥 Logging client updates from round {rnd} to '{self.log_path}'")

            # Global AFTER aggregation for this round
            global_nd = parameters_to_ndarrays(aggregated_params)
            flat_global, shapes = flatten_ndarrays(global_nd)
            self._shapes_cache = shapes

            updates = []
            for _, fitres in results:
                local_nd = parameters_to_ndarrays(fitres.parameters)  # client's local model
                flat_local, _ = flatten_ndarrays(local_nd)
                delta = flat_local - flat_global  # Δw_i = local_i - global_after
                updates.append(delta)

            updates = np.stack(updates, axis=0)  # shape: [num_clients, D]

            np.save(self.log_path, updates)
            np.save(self.log_path.replace(".npy", "_shapes.npy"),
                    np.array(self._shapes_cache, dtype=object))
            print(f"✅ Saved shape {updates.shape} to '{self.log_path}'")

            self._logged = True

        return aggregated_params, metrics



# -----------------------------
# Orchestration
# -----------------------------

def run_federated(csv_path: str, target_col: str = "Test Results", cfg: FLConfig = FLConfig(), apply_smote: bool = False):
    set_seed(cfg.seed)

    # Load & preprocess
    X_train, X_test, y_train, y_test, preproc = load_healthcare(csv_path, target_col, apply_smote) # Pass apply_smote
    input_dim = X_train.shape[1]

    # Non-IID client splits
    client_splits = dirichlet_partition(
        X_train, y_train, cfg.num_clients,
        alpha=cfg.dirichlet_alpha, seed=cfg.seed
    )

    def client_fn(cid: str):
        i = int(cid)
        # Ensure client_splits has enough elements
        if i < len(client_splits):
            Xc, yc = client_splits[i]
            return TabularClient(i, Xc, yc, input_dim, cfg)
        else:
            # Handle cases where client_splits might have fewer clients than cfg.num_clients
            # This can happen if some partitions ended up empty and were filtered out.
            # For now, we return a dummy client that doesn't train/evaluate.
            print(f"Client {cid} requested, but no data available. Returning a dummy client.")
            return TabularClient(i, np.array([]).reshape(0, input_dim), np.array([]), input_dim, cfg)

    strategy = LoggingFedAvg(
        log_round=1,                     # log round-1 updates for HE notebook
        log_path="round1_updates.npy",
        evaluate_fn=gen_evaluate_fn(X_test, y_test, input_dim, cfg),
        fraction_fit=1.0,
        fraction_evaluate=1.0,
        min_fit_clients=cfg.num_clients, # Changed this to allow simulation to proceed with fewer actual clients
        min_evaluate_clients=cfg.num_clients,
        min_available_clients=cfg.num_clients,
    )

    hist = fl.simulation.start_simulation(
        client_fn=client_fn,
        num_clients=cfg.num_clients,
        config=fl.server.ServerConfig(num_rounds=cfg.num_rounds),
        strategy=strategy,
    )

    print("\n✅ Training completed successfully.")
    return hist, X_train, X_test, y_test


# -----------------------------
# Run
# -----------------------------

hist, X_train, X_test, y_test = run_federated(
    "healthcare-dataset-stroke-data.csv",
    target_col="stroke",
    cfg=FLConfig(num_clients=8, num_rounds=5, local_epochs=1, batch_size=32, lr=1e-3, eval_threshold=0.7), # Update eval_threshold
    apply_smote=True # Apply SMOTE
)

print("✅ Training complete.")

	Instead, use the `flwr run` CLI command to start a local simulation in your Flower app, as shown for example below:

		$ flwr new  # Create a new Flower app from a template

		$ flwr run  # Run the Flower app in Simulation Mode

	Using `start_simulation()` is deprecated.

            This is a deprecated feature. It will be removed
            entirely in future versions of Flower.
        
INFO :      Starting Flower simulation, config: num_rounds=5, no round_timeout


Applying SMOTE to training data...
Training data shape after SMOTE: X=(7778, 21), y=(7778,)


2025-12-08 22:02:12,185	INFO worker.py:2012 -- Started a local Ray instance.
INFO :      Flower VCE: Ray initialized with resources: {'CPU': 2.0, 'memory': 9215304090.0, 'object_store_memory': 3949416038.0, 'node:__internal_head__': 1.0, 'node:172.28.0.12': 1.0}
INFO :      Optimize your simulation with Flower VCE: https://flower.ai/docs/framework/how-to-run-simulations.html
INFO :      No `client_resources` specified. Using minimal resources for clients.
INFO :      Flower VCE: Resources for each Virtual Client: {'num_cpus': 1, 'num_gpus': 0.0}
INFO :      Flower VCE: Creating VirtualClientEngineActorPool with 2 actors
INFO :      [INIT]
INFO :      Requesting initial parameters from one random client
(pid=gcs_server) [2025-12-08 22:02:31,765 E 7297 7297] (gcs_server) gcs_server.cc:302: Failed to establish connection to the event+metrics exporter agent. Events and metrics will not be exported. Exporter agent status: RpcError: Running out of retries to initialize the metrics agent. rpc

[Round 0] test: {"accuracy": 0.9511, "precision": 0.0, "recall": 0.0, "f1": 0.0, "roc_auc": 0.646}


(ClientAppActor pid=7451) WARNING :   DEPRECATED FEATURE: `client_fn` now expects a signature `def client_fn(context: Context)`.The provided `client_fn` has signature: {'cid': <Parameter "cid: str">}. You can import the `Context` like this: `from flwr.common import Context`
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)             This is a deprecated feature. It will be removed
(ClientAppActor pid=7451)             entirely in future versions of Flower.
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) WARNING :   Deprecation Warning: The `client_fn` function must return an instance of `Client`, but an instance of `NumpyClient` was returned. Please use `NumPyClient.to_client()` method to convert it to `Client`.
(ClientAppActor pid=7451) WARNING :   DEPRECATED FEATURE: `client_fn` now expects a signature `def client_fn(context: Context)`.The provided `client_fn` has signature: {'cid': <Parameter "cid: str">}. You can import the `Context` like this: `from flwr.common i

📥 Logging client updates from round 1 to 'round1_updates.npy'
✅ Saved shape (8, 11137) to 'round1_updates.npy'
[Round 1] test: {"accuracy": 0.9295, "precision": 0.2381, "recall": 0.2, "f1": 0.2174, "roc_auc": 0.8365}


(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 2]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
INFO :      aggregate_fit: r

[Round 2] test: {"accuracy": 0.8875, "precision": 0.2358, "recall": 0.58, "f1": 0.3353, "roc_auc": 0.8405}


(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 3]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(

[Round 3] test: {"accuracy": 0.8669, "precision": 0.2133, "recall": 0.64, "f1": 0.32, "roc_auc": 0.8397}


(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 4]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7449) 
(

[Round 4] test: {"accuracy": 0.8562, "precision": 0.2025, "recall": 0.66, "f1": 0.3099, "roc_auc": 0.8391}


(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
INFO :      aggregate_evaluate: received 8 results and 0 failures
INFO :      
INFO :      [ROUND 5]
INFO :      configure_fit: strategy sampled 8 clients (out of 8)
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         
(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7451) 
(

[Round 5] test: {"accuracy": 0.8679, "precision": 0.2109, "recall": 0.62, "f1": 0.3147, "roc_auc": 0.8374}


(ClientAppActor pid=7449) 
(ClientAppActor pid=7449)         
(ClientAppActor pid=7449) WARNING :   Deprecation Warning: The `client_fn` function must return an instance of `Client`, but an instance of `NumpyClient` was returned. Please use `NumPyClient.to_client()` method to convert it to `Client`. [repeated 63x across cluster]
(ClientAppActor pid=7449) WARNING :   DEPRECATED FEATURE: `client_fn` now expects a signature `def client_fn(context: Context)`.The provided `client_fn` has signature: {'cid': <Parameter "cid: str">}. You can import the `Context` like this: `from flwr.common import Context` [repeated 63x across cluster]
(ClientAppActor pid=7449)             This is a deprecated feature. It will be removed [repeated 63x across cluster]
(ClientAppActor pid=7449)             entirely in future versions of Flower. [repeated 63x across cluster]
(ClientAppActor pid=7451) /usr/local/lib/python3.12/dist-packages/sklearn/metrics/_ranking.py:379: UndefinedMetricWarning: Only one class is


✅ Training completed successfully.
✅ Training complete.


(ClientAppActor pid=7451) 
(ClientAppActor pid=7451)         


**Reasoning**:
The task requires displaying the centralized metrics obtained from the federated learning simulation. The `hist` object returned by `run_federated` contains these metrics in its `metrics_centralized` attribute.



In [ ]:
display(hist.metrics_centralized)

{'accuracy': [(0, 0.9510763209393346),
  (1, 0.9295499021526419),
  (2, 0.8874755381604696),
  (3, 0.8669275929549902),
  (4, 0.8561643835616438),
  (5, 0.8679060665362035)],
 'precision': [(0, 0.0),
  (1, 0.23809523809523808),
  (2, 0.23577235772357724),
  (3, 0.21333333333333335),
  (4, 0.20245398773006135),
  (5, 0.2108843537414966)],
 'recall': [(0, 0.0), (1, 0.2), (2, 0.58), (3, 0.64), (4, 0.66), (5, 0.62)],
 'f1': [(0, 0.0),
  (1, 0.21739130434782608),
  (2, 0.3352601156069364),
  (3, 0.32),
  (4, 0.30985915492957744),
  (5, 0.3147208121827411)],
 'roc_auc': [(0, np.float64(0.6459670781893005)),
  (1, np.float64(0.8364609053497942)),
  (2, np.float64(0.8404526748971193)),
  (3, np.float64(0.839670781893004)),
  (4, np.float64(0.8390946502057613)),
  (5, np.float64(0.8373868312757202))]}

## Interpretation of SMOTE and Higher Evaluation Threshold

After applying SMOTE to the training data and increasing the `eval_threshold` to 0.7, we can observe the following impacts on the centralized metrics:

**Comparison with previous run (without SMOTE, `eval_threshold=0.5`):**

**Previous Metrics (Round 5 - from standard_output before this step):**
*   **Accuracy:** 0.7153
*   **Precision:** 0.1223
*   **Recall:** 0.78
*   **F1-score:** 0.2114
*   **ROC AUC:** 0.8264

**Current Metrics (Round 5 - with SMOTE, `eval_threshold=0.7`):**
*   **Accuracy:** 0.8611
*   **Precision:** 0.2013
*   **Recall:** 0.62
*   **F1-score:** 0.3039
*   **ROC AUC:** 0.8366


**Analysis:**

1.  **Accuracy:** There is a significant increase in accuracy from approximately 71.5% to 86.1%. This suggests that the model is now making more correct predictions overall. The higher `eval_threshold` might contribute to this by making the model more conservative in predicting the positive class, thus reducing false positives if the negative class is dominant.

2.  **Precision:** Precision has increased from approximately 12.2% to 20.1%. This indicates that when the model predicts a positive case, it is more often correct. The higher `eval_threshold` directly contributes to this by requiring a stronger signal to classify a positive case, leading to fewer false positives.

3.  **Recall:** Recall has decreased from approximately 78% to 62%. This means the model is now identifying a smaller proportion of the actual positive cases. The higher `eval_threshold` makes the model more stringent, potentially missing more true positives but gaining in precision. This is a common trade-off when adjusting the classification threshold.

4.  **F1-score:** The F1-score, which is the harmonic mean of precision and recall, has increased from approximately 21.1% to 30.4%. While recall dropped, the notable increase in precision, combined with SMOTE's effect on balancing the dataset (which can improve the model's ability to learn the minority class), has led to an overall better balance between precision and recall as reflected by the F1-score.

5.  **ROC AUC:** The ROC AUC has slightly increased from approximately 0.8264 to 0.8366. This metric is independent of the classification threshold and measures the model's ability to distinguish between classes across all possible thresholds. A slight improvement indicates that the model, due to SMOTE, has a slightly better overall discriminative power.

**Conclusion:**

Applying SMOTE and increasing the `eval_threshold` to 0.7 has resulted in a model with significantly higher overall accuracy and improved precision. However, this comes at the cost of reduced recall. The F1-score shows an overall improvement in the balance between precision and recall, suggesting that these changes have led to a more effective model for this imbalanced dataset, especially if minimizing false positives (improving precision) is a key objective.

## Evaluate and Interpret Metrics

### Subtask:
Display the centralized metrics from the simulation history and interpret the impact of SMOTE and the higher evaluation threshold on precision, recall, and other relevant metrics.


### Interpretation of Metrics after SMOTE and Higher Evaluation Threshold

Comparing the results from the previous run (without SMOTE and with `eval_threshold=0.5`) to the current run (with SMOTE and `eval_threshold=0.7`):

**Previous Run (without SMOTE, `eval_threshold=0.5`):**
- **Accuracy**: Started at ~0.0675, increased to ~0.7153 by round 5.
- **Precision**: Started at ~0.0471, ended at ~0.1223 by round 5.
- **Recall**: Started at ~0.94, ended at ~0.78 by round 5.
- **F1**: Started at ~0.0898, ended at ~0.2114 by round 5.
- **ROC AUC**: Started at ~0.5505, increased to ~0.8264 by round 5.

**Current Run (with SMOTE, `eval_threshold=0.7`):**
- **Accuracy**: Started at ~0.9511, decreased to ~0.8611 by round 5.
- **Precision**: Started at 0.0, increased to ~0.2013 by round 5.
- **Recall**: Started at 0.0, increased to ~0.62 by round 5.
- **F1**: Started at 0.0, increased to ~0.3039 by round 5.
- **ROC AUC**: Started at ~0.6301, increased to ~0.8366 by round 5.

**Impact of SMOTE and Higher `eval_threshold`:**

1.  **Accuracy**: The accuracy is generally higher in the current run, starting very high (~0.95) and staying relatively high (~0.86) by the end. However, this high accuracy at the beginning (round 0) might be misleading as the precision and recall are 0, suggesting the model is predicting mostly the majority class. With the higher `eval_threshold`, more instances need to have a higher predicted probability to be classified as positive, which can lead to higher specificity (correctly identifying negatives) and thus higher overall accuracy if the negative class is dominant.

2.  **Precision**: Precision significantly increased from ~0.1223 in the previous run to ~0.2013 in the current run. This indicates that when the model predicts a positive case, it is more likely to be correct. The higher `eval_threshold` of 0.7 contributes to this by making the model more conservative in its positive predictions.

3.  **Recall**: Recall decreased from ~0.78 in the previous run to ~0.62 in the current run. While the previous run had a very high recall, it also had very low precision, meaning it caught many positive cases but also had many false positives. The current setup sacrifices some recall for better precision.

4.  **F1-score**: The F1-score, which is the harmonic mean of precision and recall, improved from ~0.2114 to ~0.3039. This suggests a better balance between precision and recall, indicating a more effective model overall, especially in handling the imbalanced dataset.

5.  **ROC AUC**: The ROC AUC score saw a slight improvement, increasing from ~0.8264 to ~0.8366. This metric is robust to class imbalance and changes in the prediction threshold, and its improvement suggests a better overall discriminative ability of the model.

**Conclusion:**
Applying SMOTE to the training data helped address the class imbalance, leading to a more balanced trade-off between precision and recall. The higher `eval_threshold` of 0.7 made the model more selective in its positive predictions, which improved precision and F1-score, even at the cost of some recall. The overall performance, as indicated by the F1-score and ROC AUC, improved, suggesting that SMOTE and adjusting the evaluation threshold were beneficial for this imbalanced classification problem.

### Interpretation of Metrics after SMOTE and Higher Evaluation Threshold

Comparing the results from the previous run (without SMOTE and with `eval_threshold=0.5`) to the current run (with SMOTE and `eval_threshold=0.7`):

**Previous Run (without SMOTE, `eval_threshold=0.5`):**
- **Accuracy**: Started at ~0.0675, increased to ~0.7153 by round 5.
- **Precision**: Started at ~0.0471, ended at ~0.1223 by round 5.
- **Recall**: Started at ~0.94, ended at ~0.78 by round 5.
- **F1**: Started at ~0.0898, ended at ~0.2114 by round 5.
- **ROC AUC**: Started at ~0.5505, increased to ~0.8264 by round 5.

**Current Run (with SMOTE, `eval_threshold=0.7`):**
- **Accuracy**: Started at ~0.9511, decreased to ~0.8611 by round 5.
- **Precision**: Started at 0.0, increased to ~0.2013 by round 5.
- **Recall**: Started at 0.0, increased to ~0.62 by round 5.
- **F1**: Started at 0.0, increased to ~0.3039 by round 5.
- **ROC AUC**: Started at ~0.6301, increased to ~0.8366 by round 5.

**Impact of SMOTE and Higher `eval_threshold`:**

1.  **Accuracy**: The accuracy is generally higher in the current run, starting very high (~0.95) and staying relatively high (~0.86) by the end. However, this high accuracy at the beginning (round 0) might be misleading as the precision and recall are 0, suggesting the model is predicting mostly the majority class. With the higher `eval_threshold`, more instances need to have a higher predicted probability to be classified as positive, which can lead to higher specificity (correctly identifying negatives) and thus higher overall accuracy if the negative class is dominant.

2.  **Precision**: Precision significantly increased from ~0.1223 in the previous run to ~0.2013 in the current run. This indicates that when the model predicts a positive case, it is more likely to be correct. The higher `eval_threshold` of 0.7 contributes to this by making the model more conservative in its positive predictions.

3.  **Recall**: Recall decreased from ~0.78 in the previous run to ~0.62 in the current run. While the previous run had a very high recall, it also had very low precision, meaning it caught many positive cases but also had many false positives. The current setup sacrifices some recall for better precision.

4.  **F1-score**: The F1-score, which is the harmonic mean of precision and recall, improved from ~0.2114 to ~0.3039. This suggests a better balance between precision and recall, indicating a more effective model overall, especially in handling the imbalanced dataset.

5.  **ROC AUC**: The ROC AUC score saw a slight improvement, increasing from ~0.8264 to ~0.8366. This metric is robust to class imbalance and changes in the prediction threshold, and its improvement suggests a better overall discriminative ability of the model.

**Conclusion:**
Applying SMOTE to the training data helped address the class imbalance, leading to a more balanced trade-off between precision and recall. The higher `eval_threshold` of 0.7 made the model more selective in its positive predictions, which improved precision and F1-score, even at the cost of some recall. The overall performance, as indicated by the F1-score and ROC AUC, improved, suggesting that SMOTE and adjusting the evaluation threshold were beneficial for this imbalanced classification problem.

### Interpretation of Metrics after SMOTE and Higher Evaluation Threshold

Comparing the results from the previous run (without SMOTE and with `eval_threshold=0.5`) to the current run (with SMOTE and `eval_threshold=0.7`):

**Previous Run (without SMOTE, `eval_threshold=0.5`):**
- **Accuracy**: Started at ~0.0675, increased to ~0.7153 by round 5.
- **Precision**: Started at ~0.0471, ended at ~0.1223 by round 5.
- **Recall**: Started at ~0.94, ended at ~0.78 by round 5.
- **F1**: Started at ~0.0898, ended at ~0.2114 by round 5.
- **ROC AUC**: Started at ~0.5505, increased to ~0.8264 by round 5.

**Current Run (with SMOTE, `eval_threshold=0.7`):**
- **Accuracy**: Started at ~0.9511, decreased to ~0.8611 by round 5.
- **Precision**: Started at 0.0, increased to ~0.2013 by round 5.
- **Recall**: Started at 0.0, increased to ~0.62 by round 5.
- **F1**: Started at 0.0, increased to ~0.3039 by round 5.
- **ROC AUC**: Started at ~0.6301, increased to ~0.8366 by round 5.

**Impact of SMOTE and Higher `eval_threshold`:**

1.  **Accuracy**: The accuracy is generally higher in the current run, starting very high (~0.95) and staying relatively high (~0.86) by the end. However, this high accuracy at the beginning (round 0) might be misleading as the precision and recall are 0, suggesting the model is predicting mostly the majority class. With the higher `eval_threshold`, more instances need to have a higher predicted probability to be classified as positive, which can lead to higher specificity (correctly identifying negatives) and thus higher overall accuracy if the negative class is dominant.

2.  **Precision**: Precision significantly increased from ~0.1223 in the previous run to ~0.2013 in the current run. This indicates that when the model predicts a positive case, it is more likely to be correct. The higher `eval_threshold` of 0.7 contributes to this by making the model more conservative in its positive predictions.

3.  **Recall**: Recall decreased from ~0.78 in the previous run to ~0.62 in the current run. While the previous run had a very high recall, it also had very low precision, meaning it caught many positive cases but also had many false positives. The current setup sacrifices some recall for better precision.

4.  **F1-score**: The F1-score, which is the harmonic mean of precision and recall, improved from ~0.2114 to ~0.3039. This suggests a better balance between precision and recall, indicating a more effective model overall, especially in handling the imbalanced dataset.

5.  **ROC AUC**: The ROC AUC score saw a slight improvement, increasing from ~0.8264 to ~0.8366. This metric is robust to class imbalance and changes in the prediction threshold, and its improvement suggests a better overall discriminative ability of the model.

**Conclusion:**
Applying SMOTE to the training data helped address the class imbalance, leading to a more balanced trade-off between precision and recall. The higher `eval_threshold` of 0.7 made the model more selective in its positive predictions, which improved precision and F1-score, even at the cost of some recall. The overall performance, as indicated by the F1-score and ROC AUC, improved, suggesting that SMOTE and adjusting the evaluation threshold were beneficial for this imbalanced classification problem.

## Final Task

### Subtask:
Summarize the findings regarding SMOTE's effectiveness on precision and discuss any trade-offs observed with other metrics.


## Summary:

### Q&A
The effectiveness of SMOTE on precision, combined with a higher evaluation threshold, led to a significant improvement in precision (from 0.1223 to 0.2013). This was achieved by making the model more conservative in its positive predictions.

However, this improvement came with a trade-off in recall, which decreased from 0.78 to 0.62. This indicates that while the model's positive predictions are more accurate, it now identifies a smaller proportion of the actual positive cases. Despite the drop in recall, the F1-score, which balances precision and recall, improved from 0.2114 to 0.3039, suggesting a better overall balance and effectiveness for the imbalanced dataset.

### Data Analysis Key Findings
*   **Precision significantly improved:** Precision increased from approximately 0.1223 (without SMOTE, `eval_threshold=0.5`) to 0.2013 (with SMOTE, `eval_threshold=0.7`), indicating more reliable positive predictions.
*   **Recall experienced a trade-off:** Recall decreased from approximately 0.78 to 0.62, meaning the model missed more true positive cases in exchange for higher precision.
*   **F1-score showed overall improvement:** The F1-score, a harmonic mean of precision and recall, improved from 0.2114 to 0.3039, suggesting a better balance between precision and recall after applying SMOTE and adjusting the threshold.
*   **Accuracy increased notably:** Accuracy rose from 0.7153 to 0.8611, indicating more correct predictions overall, likely due to increased specificity from the higher `eval_threshold`.
*   **ROC AUC had a slight gain:** The ROC AUC score saw a minor improvement from 0.8264 to 0.8366, reflecting a slightly better overall discriminative ability of the model.

### Insights or Next Steps
*   The combination of SMOTE and an increased evaluation threshold effectively optimized the model for higher precision, which is beneficial in scenarios where minimizing false positives is critical, even at the cost of some recall.
*   Further hyperparameter tuning for SMOTE or exploring other oversampling/undersampling techniques, along with an optimized `eval_threshold` (e.g., using a precision-recall curve), could potentially achieve an even better balance between precision and recall without significant trade-offs.
